src/v/cluster/namespaced_cache.h

```diff
@@ -0,0 +1,464 @@
+    {
+        cache.evict(pre_eviction_hook, post_eviction_hook)
+    } -> std::same_as<bool>;
+    { cache.remove(entry) } -> std::same_as<void>;
+    { cache.touch(entry) } -> std::same_as<void>;
+    { cache.size() } -> std::same_as<size_t>;
+};
+
+/**
+ * Simple cache with LRU eviction policy
+ */
+template<
+  typename EntryT,
+  safe_intrusive_list_hook EntryT::*HookPtr,
+  pre_eviction_hook<EntryT> PreEvictionHookT,
+  post_eviction_hook<EntryT> PostEvictionHookT>
+struct lru_cache {
+    using list_type = uncounted_intrusive_list<EntryT, HookPtr>;
+
+    void insert(EntryT&);
+    bool evict(const PreEvictionHookT&, const PostEvictionHookT&);
+    void remove(const EntryT&);
+    void touch(EntryT&);
+    size_t size() const { return _size; }
+
+private:
+    size_t _size{0};
+    list_type _lru;
+};
+
+/**
+ * A simple cache that maintains an LRU per namespace.
+ *
+ * The cache allows to maintain per namespace lru cache with fair distribution
+ * of available cache slots across all namespaces. Cache size is
+ * controlled by setting maximum number of entries. If adding new namespace to
+ * the cache would result in other namespace to have less than
+ * min_slots_per_namespace cache slots an exception is thrown.
+ *
+ * Maximum number of namespaces that cache can handle can be calculated as:
+ *
+ *  floor(_max_size/_min_slots_per_namespace)
+ *
+ * With cache PreEvictionHookT template parameter it is possible to prevent some
+ * entries from being evicted. A predicate should return 'false' if an entry is
+ * not evictable.
+ *
+ * Eviction policy can be controlled with the CacheT parameter. The CacheT
+ * parameter allow to tune per namespace caching policy.
+ *
+ * For now the only provided eviction policy is a simple LRU. In future we may
+ * experience with different caching strategies like f.e. SIEVE or S3-FIFO.
+ *
+ */
+template<
+  typename EntryT,
+  typename NamespaceT,
+  safe_intrusive_list_hook EntryT::*HookPtr,
+  pre_eviction_hook<EntryT> PreEvictionHookT = always_allow_to_evict,
+  post_eviction_hook<EntryT> PostEvictionHookT = noop_post_eviction_hook,
+  namespace_cache<EntryT, HookPtr, PreEvictionHookT, PostEvictionHookT> CacheT
+  = lru_cache<EntryT, HookPtr, PreEvictionHookT, PostEvictionHookT>>
+class namespaced_cache {
+private:
+    using namespaces_t
+      = absl::node_hash_map<NamespaceT, std::unique_ptr<CacheT>>;
+
+public:
+    struct stats {
+        size_t total_size{0};
+        absl::node_hash_map<NamespaceT, size_t> namespaces;
+    };
+
+    /**
+     * Creates a namespaced cache with desired maximum size and minimal allowed
+     * number of slots per namespace
+     */
+    explicit namespaced_cache(
+      config::binding<size_t> total_max_size,
+      config::binding<size_t> min_slots_per_namespace,
+      const PreEvictionHookT& pre_eviction_hook = PreEvictionHookT(),
+      const PostEvictionHookT& post_eviction_hook = PostEvictionHookT())
+      : _max_size(std::move(total_max_size))
```

Comment:
carrying over comment from original PR, updating these bindings may have to reset the cache to evict any excess entries.

Reply:
done